controllers/cloudstackfailuredomain_controller.go

```diff
@@ -0,0 +1,131 @@
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackfailuredomains/finalizers,verbs=update
+
+// CloudStackFailureDomainReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackFailureDomains.
+// The runner does the actual reconciliation.
+type CloudStackFailureDomainReconciliationRunner struct {
+	*csCtrlrUtils.ReconciliationRunner
+	ReconciliationSubject *infrav1.CloudStackFailureDomain
+	IsoNet                *infrav1.CloudStackIsolatedNetwork
+}
+
+// Initialize a new CloudStackFailureDomain reconciliation runner with concrete types and initialized member fields.
+func NewCSFailureDomainReconciliationRunner() *CloudStackFailureDomainReconciliationRunner {
+	// Set concrete type and init pointers.
+	r := &CloudStackFailureDomainReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackFailureDomain{}}
+	r.IsoNet = &infrav1.CloudStackIsolatedNetwork{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	r.ReconciliationRunner = csCtrlrUtils.NewRunner(r, r.ReconciliationSubject, "CloudStackFailureDomain")
+
+	return r
+}
+
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CloudStackFailureDomainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	return NewCSFailureDomainReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
+}
+
+// Reconcile on the ReconciliationRunner actually attempts to modify or create the reconciliation subject.
+func (r *CloudStackFailureDomainReconciliationRunner) Reconcile() (retRes ctrl.Result, retErr error) {
+	res, err := r.AsFailureDomainUser(&r.ReconciliationSubject.Spec)()
+	if r.ShouldReturn(res, err) {
+		return res, err
+	}
+	// Prevent premature deletion.
+	controllerutil.AddFinalizer(r.ReconciliationSubject, infrav1.FailureDomainFinalizer)
+
+	// Start by purely data fetching information about the zone and specified network.
+	if err := r.CSUser.ResolveZone(&r.ReconciliationSubject.Spec.Zone); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "resolving CloudStack zone information")
+	}
+	if err := r.CSUser.ResolveNetworkForZone(&r.ReconciliationSubject.Spec.Zone); err != nil &&
+		!csCtrlrUtils.ContainsNoMatchSubstring(err) {
+		return ctrl.Result{}, errors.Wrap(err, "resolving Cloudstack network information")
+	}
+
+	// Check if the passed network was an isolated network or the network was missing. In either case, create a
+	// CloudStackIsolatedNetwork to manage the many intricacies and wait until CloudStackIsolatedNetwork is ready.
+	if r.ReconciliationSubject.Spec.Zone.Network.ID == "" ||
+		r.ReconciliationSubject.Spec.Zone.Network.Type == infrav1.NetworkTypeIsolated {
+		netName := r.ReconciliationSubject.Spec.Zone.Network.Name
```

Comment:
could you offset the spacing on the `if` condition here a bit? It blends with the main logic and makes it a bit confusing at first glance